File: solutions/cpp/secret-handshake/1/secret_handshake.cpp
#include "secret_handshake.h"

#include <algorithm>
#include <bitset>
#include <utility>

namespace secret_handshake
{
    using handshake_event_t = std::pair<std::uint8_t, std::string>;
    
    const std::vector<handshake_event_t> handshake_events = {
        {0, "wink"},
        {1, "double blink"},
        {2, "close your eyes"},
        {3, "jump"},
        {4, "reverse"}};

    std::vector<std::string>
    commands(std::uint8_t handshake_int)
    {
        std::bitset<5> handshake_bitset{handshake_int};
        std::vector<std::string> handshake;

        for (auto handshake_event : handshake_events)
            if (handshake_bitset[handshake_event.first])
            {
                if (handshake_event.second == "reverse")
                {
                    std::reverse(handshake.begin(), handshake.end());
                }
                else
                {
                    handshake.push_back(handshake_event.second);
                }
            }

        return handshake;
    }

} // namespace secret_handshake

File: solutions/haskell/secret-handshake/1/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits

handshake :: Int -> [String]
handshake n
  | n .&. 0x10 == 0x10 = reverse (handshake $ n `xor` 0x10) -- Must be outmost call
  | n .&. 0x1 == 0x1 = "wink" : (handshake $ n `xor` 0x1)
  | n .&. 0x2 == 0x2 = "double blink" : (handshake $ n `xor` 0x2)
  | n .&. 0x4 == 0x4 = "close your eyes" : (handshake $ n `xor` 0x4)
  | n .&. 0x8 == 0x8 = "jump" : (handshake $ n `xor` 0x8)
  | otherwise = []
